spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/SetRequestHostHeaderGatewayFilterFactory.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gateway.filter.factory;
+
+import org.springframework.http.server.reactive.ServerHttpRequest;
+import reactor.core.publisher.Mono;
+
+import org.springframework.cloud.gateway.filter.GatewayFilter;
+import org.springframework.cloud.gateway.filter.GatewayFilterChain;
+import org.springframework.cloud.gateway.support.ServerWebExchangeUtils;
+import org.springframework.web.server.ServerWebExchange;
+
+import static org.springframework.cloud.gateway.support.GatewayToStringStyler.filterToStringCreator;
+
+/**
+ * @author Andrew Fitzgerald
+ */
+public class SetRequestHostHeaderGatewayFilterFactory
+		extends AbstractGatewayFilterFactory<SetRequestHostHeaderGatewayFilterFactory.Config> {
+
+
+	public SetRequestHostHeaderGatewayFilterFactory() {
+		super(Config.class);
+	}
+
+	@Override
+	public GatewayFilter apply(Config config) {
+		return new GatewayFilter() {
+			@Override
+			public Mono<Void> filter(ServerWebExchange exchange,
+					GatewayFilterChain chain) {
+				String value = ServerWebExchangeUtils.expand(exchange, config.getHost());
+
+				ServerHttpRequest request = exchange.getRequest().mutate()
+						.headers(httpHeaders -> {
+							httpHeaders.remove("Host");
+							httpHeaders.add("Host", value);
+						})
+						.build();
+
+				return chain.filter(exchange.mutate().request(request).build());
+			}
+
+			@Override
+			public String toString() {
+				return filterToStringCreator(SetRequestHostHeaderGatewayFilterFactory.this)
+				.append(config.getHost()).toString();
+			}
+		};
+	}
+
+	public static class Config {
```

Comment:
There is a `NameConfig` class that is single value, but there is precedent (see `PrefixPathGatewayFilterFactory`) to have the config key have meaning, so I say keep this.